cmd/bosun/conf/rule/modify.go

```diff
@@ -0,0 +1,173 @@
+	}
+	if currentDiff != diff {
+		return fmt.Errorf("couldn't save config file because the change and supplied diff do not match the current diff")
+	}
+	if err = c.SaveConf(newConf); err != nil {
+		return fmt.Errorf("couldn't save config file: %v", err)
+	}
+	if c.saveHook != nil {
+		err := c.callSaveHook(c.Name, user, message, args...)
+		if err != nil {
+			sErr := c.SaveConf(c)
+			restore := "successful"
+			if sErr != nil {
+				restore = sErr.Error()
+			}
+			return fmt.Errorf("failed to call save hook: %v. Restoring config: %v", err, restore)
+		}
+	}
+	err = c.reload()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// BulkEdit applies sequental edits to the configuration file. Each individual edit
+// must generate a valid configuration or the edit request will fail.
+func (c *Conf) BulkEdit(edits conf.BulkEditRequest) error {
+	select {
+	case c.writeLock <- true:
+		// Got Write Lock
+	default:
+		return fmt.Errorf("cannot write alert, write in progress")
+	}
+	defer func() {
+		<-c.writeLock
+	}()
+	newConf := c
+	var err error
+	for _, edit := range edits {
+		var l Location
+		switch edit.Type {
+		case "alert":
+			a := newConf.GetAlert(edit.Name)
+			if a != nil {
+				l = a.Locator.(Location)
+			}
+		case "template":
+			t := newConf.GetTemplate(edit.Name)
+			if t != nil {
+				l = t.Locator.(Location)
+			}
+		case "notification":
+			n := newConf.GetNotification(edit.Name)
+			if n != nil {
+				l = n.Locator.(Location)
+			}
+		case "lookup":
+			look := newConf.GetLookup(edit.Name)
+			if look != nil {
+				l = look.Locator.(Location)
+			}
+		case "macro":
+			m := newConf.GetMacro(edit.Name)
+			if m != nil {
+				l = m.Locator.(Location)
+			}
+		default:
+			return fmt.Errorf("%v is an unsuported type for bulk edit. must be alert, template, notification, lookup or macro", edit.Type)
+		}
+		var rawConf string
+		if edit.Delete {
+			if l == nil {
+				return fmt.Errorf("could not delete %v:%v - not found", edit.Type, edit.Name)
+			}
+			rawConf = removeSection(l, newConf.RawText)
+		} else {
+			rawConf = writeSection(l, newConf.RawText, edit.Text)
+		}
+		newConf, err = NewConf(c.Name, c.backends, rawConf)
+		if err != nil {
+			return fmt.Errorf("could not create new conf: failed on step %v:%v : %v", edit.Type, edit.Name, err)
+		}
+	}
+	if err := c.SaveConf(newConf); err != nil {
+		return fmt.Errorf("couldn't save config file: %v", err)
+	}
+	err = c.reload()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Location stores the start byte position and end byte position of
+// an object in the raw configuration
+type Location []int
+
+func writeSection(l Location, orginalRaw, newText string) string {
+	var newRawConf bytes.Buffer
+	if l == nil {
+		newRawConf.WriteString(orginalRaw)
+		newRawConf.WriteString("\n")
+		newRawConf.WriteString(newText)
+		newRawConf.WriteString("\n")
+		return newRawConf.String()
+	}
+	newRawConf.WriteString(orginalRaw[:getLocationStart(l)])
```

Comment:
I love the emoji, but these incantations are repeated a few times. How about a helper on the Location type?

`textBefore(originalRaw, l)`
`textAfter(originalRaw, l)`

may not be worth it.